src/src/PluginStructs/P014_data_struct.cpp

```diff
@@ -175,148 +272,281 @@ uint8_t P014_data_struct::checkCRC(uint16_t data, uint8_t check)
+  chip_id = sernum_b; //we save the chip identifier
+  //Wire.read(); //ignore CRC
+  sernum_b <<= 8;
+  sernum_b |= Wire.read(); //SNB_2
+  Wire.read(); //ignore CRC
+  sernum_b <<= 8;
+  sernum_b |= Wire.read(); //SNB_1
+  //Wire.read(); //ignore CRC
+  sernum_b <<= 8;
+  sernum_b |= Wire.read(); //SNB_0
+  Wire.read(); //ignore CRC
+
+
+  String log = F("SI7013 : sn=");
+  log += String(sernum_a,HEX);
+  log += F("-");
+  log += String(sernum_b,HEX);
+  addLog(LOG_LEVEL_INFO,log);
+
+  return 0; //ok
+}
+
+
+bool P014_data_struct::enablePowerForADC(uint8_t i2caddr){
+  //set VOUT
+  //Get the current register value
+
+  //we know the type of circuit used based on the i2c address 
 
-  // Check CRC of data received
-  if (checkCRC(raw, checksum) != 0) {
-    addLog(LOG_LEVEL_ERROR, F("SI7021 : checksum error!"));
-    return -1;
+
+  bool ok=false;
+  uint8_t reg = I2C_read8_reg(i2caddr,SI7013_READ_REG2, &ok);
+
+  if (!ok) {
+    addLog(LOG_LEVEL_ERROR, F("SI7013: Could not read REG2!"));
+    return false;
+  }
+    
+  
+  if (i2caddr == SI7013_I2C_ADDRESS_AD0_1){
+    ok = I2C_write8_reg(i2caddr,SI7013_WRITE_REG2, (reg & B11111000) | (2+4+64) );//set last three bits (VIN bufered, Vref=VDD, VOUT=GND) and No-Hold for bit 6
+  }else{
+    ok = I2C_write8_reg(i2caddr,SI7013_WRITE_REG2,reg | (1+2+4+64) );//set last three bits to 1 (VIN bufered, Vref=VDD, VOUT=VDD) and No-Hold for bit 6
+  }
+  if (!ok){
+    addLog(LOG_LEVEL_ERROR, F("SI7013: Could not write REG2!"));
+    return false;
   }
 
-  // Humidity
-  if ((datatype == SI7021_MEASURE_HUM) || (datatype == SI7021_MEASURE_HUM_HM)) {
-    // Convert value to Himidity percent
-    // pm-cz: it is possible to enable decimal places for humidity as well by multiplying the value in formula by 100
-    data = ((1250 * (long)raw) >> 16) - 60;
+  return true;
+}
 
-    // Datasheet says doing this check
-    if (data > 1000) { data = 1000; }
+bool P014_data_struct::disablePowerForADC(uint8_t i2caddr){
+  //set VOUT
+  //Get the current register value
 
-    if (data < 0) { data = 0; }
+  //we know the type of circuit used based on the i2c address 
 
-    // pm-cz: Let us make sure we have enough precision due to ADC bits
-    if (resolution == SI7021_RESOLUTION_12T_08RH) {
-      data  = (data + 5) / 10;
-      data *= 10;
+
+  bool ok=false;
+  uint8_t reg = I2C_read8_reg(i2caddr,SI7013_READ_REG2, &ok);
+
+  if (!ok) {
+    addLog(LOG_LEVEL_ERROR, F("SI7013: Could not read REG2!"));
+    return false;
+  }
+    
+  
+  if (i2caddr == SI7013_I2C_ADDRESS_AD0_1){
+    ok = I2C_write8_reg(i2caddr,SI7013_WRITE_REG2,reg | (1+2+4+64) );//set last three bits to 1 (VIN bufered, Vref=VDD, VOUT=VDD) and No-Hold for bit 6
+  }else{
+    ok = I2C_write8_reg(i2caddr,SI7013_WRITE_REG2, (reg & B11111000) | (2+4+64) );//set last three bits (VIN bufered, Vref=VDD, VOUT=GND) and No-Hold for bit 6
+  }
+  if (!ok){
+    addLog(LOG_LEVEL_ERROR, F("SI7013: Could not write REG2!"));
```

Comment:
Just FYI, you don't need to check this as there is no allocation done if the loglevel is not active.

Reply:
not sure I understand the issue here. Can you be more explicit please?

Reply:
As I pointed out for any other logging call, that you need to check if the loglevel is enabled before creating the log string.
However, calling `addLog(...., F(""))` like here, you don't need to add such a check as it will be checked before there is any memory allocated.
So it was just to help you better understand the code.

TL;DR no need to change here, just FYI ;)